src/codegen/unwinding.cpp

```diff
@@ -410,23 +417,83 @@ static unw_word_t getFunctionEnd(unw_word_t ip) {
     return pip.end_ip;
 }
 
+static bool inASTInterpreterExecuteInner(unw_word_t ip) {
+    static unw_word_t interpreter_instr_end = getFunctionEnd((unw_word_t)interpreter_instr_addr);
+    return ((unw_word_t)interpreter_instr_addr <= ip && ip < interpreter_instr_end);
+}
+
+static bool inGeneratorEntry(unw_word_t ip) {
+    static unw_word_t generator_entry_end = getFunctionEnd((unw_word_t)generatorEntry);
+    return ((unw_word_t)generatorEntry <= ip && ip < generator_entry_end);
+}
+
+
+static inline unw_word_t get_cursor_reg(unw_cursor_t* cursor, int reg) {
+    unw_word_t v;
+    unw_get_reg(cursor, reg, &v);
+    return v;
+}
+static inline unw_word_t get_cursor_ip(unw_cursor_t* cursor) {
+    return get_cursor_reg(cursor, UNW_REG_IP) - 1;
+}
+static inline unw_word_t get_cursor_bp(unw_cursor_t* cursor) {
+    return get_cursor_reg(cursor, UNW_TDEP_BP);
+}
+
+template <typename FrameFunc>
+bool unwindProcessFrame(unw_word_t ip, unw_word_t bp, unw_cursor_t* cursor, FrameFunc func) {
+    CompiledFunction* cf = getCFForAddress(ip);
+    bool jitted = cf != NULL;
+    if (!cf) {
+        if (inASTInterpreterExecuteInner(ip)) {
+            cf = getCFForInterpretedFrame((void*)bp);
+            assert(cf);
+        }
+    }
+
+    if (!cf)
+        return false;
+
+    PythonFrameIteratorImpl info(jitted ? PythonFrameId::COMPILED : PythonFrameId::INTERPRETED, ip, bp, cf);
+    if (jitted) {
+        // Try getting all the callee-save registers, and save the ones we were able to get.
+        // Some of them may be inaccessible, I think because they weren't defined by that
+        // stack frame, which can show up as a -UNW_EBADREG return code.
+        for (int i = 0; i < 16; i++) {
+            if (!assembler::Register::fromDwarf(i).isCalleeSave())
+                continue;
+            unw_word_t r;
+            int code = unw_get_reg(cursor, i, &r);
+            ASSERT(code == 0 || code == -UNW_EBADREG, "%d %d", code, i);
+            if (code == 0) {
+                info.regs[i] = r;
+                info.regs_valid |= (1 << i);
+            }
+        }
+    }
+
+    if (threading::ThreadStateInternal::getUnwindState() == UNWIND_STATE_NORMAL) {
```

Comment:
I've moved the code for getting/setting the unwind state out of `unwindProcessFrame`, and while it still works (and is better from a separation of concerns standpoint), it's a lot of c&p code doing the same checks.  I came up with the following higher order template func which seems to help.  It's only used in unwindPythonStack and maybeTracebackHere

https://gist.github.com/toshok/abf78446ce2adb59782d

that said, the checks are still performed since we go through unwindPythonStack.  It should be impossible to be in the RERAISE state, though, since the first thing unwindPythonStack does is set unwind_state to NORMAL.